### PR TITLE
[api] inject fail points to each endpoint

### PR DESCRIPTION
**Pull request by xli**
#9193 

API failpoint name pattern: "api::endpoint_<endpoint name>"

Test: 

```
cargo build -p diem-node --all-features
```

update node configuration to enable failpoints:
```
failpoints:
  api::endpoint_index: 100%return
```

Curl the index endpoint should return 500 error
```
curl http://localhost:8080/
{"code":500,"message":"unexpected internal error for endpoint_index"}
```

### CODE REVIEW

File: api/src/failpoint.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+#[allow(unused_imports)]
+use anyhow::{format_err, Result};
+use diem_api_types::Error;
+
+#[allow(unused_variables)]
+#[inline]
+pub fn fail_point(name: &str) -> Result<(), Error> {
+    Ok(fail::fail_point!(format!("api::{}", name).as_str(), |_| {
+        Err(format_err!("unexpected internal error for {}", name).into())
```
Comment (dimroc) [inline api/src/failpoint.rs:12]: Worth highlighting the purpose of this function is to help `fail_point!` know the type of error to return in the presence of failure. Documented here: https://docs.rs/fail/latest/fail/#early-return

**Review by dimroc — CHANGES_REQUESTED** [review]
Having one test to ensure that `failpoint.rs` works as expected would give us a lot more confidence.

https://docs.rs/fail/latest/fail/#usage-in-tests

**Comment by xli** [comment]
> Having one test to ensure that `failpoint.rs` works as expected would give us a lot more confidence.
> 
> https://docs.rs/fail/latest/fail/#usage-in-tests

We will need create a fail-point-specific test binary for failpoint test, because failpoint is global resource, from the above doc:
> Even if a test does not itself turn on any fail points, code that it runs could trigger a fail point that was configured by another thread.

I think compiler level check that failpoint related code is good enough for this case.

File: api/src/failpoint.rs (hunk 1 above)
Comment (xli) [reply api/src/failpoint.rs:12]: are you asking for adding document?

Comment (xli) [reply api/src/failpoint.rs:12]: are you asking for adding document?

**Comment by dimroc** [comment]
I envisioned a rust test as show below, but didn't realize it had that much overhead. Based on https://fb.workplace.com/groups/2044217752540140/permalink/2389820681313177/ it seems that we have significant test infra in place. Will approve to unblock but still feel a small rust unit test would be helpful.

```rust
#[test]
#[should_panic]
fn test_fallible_work() {
    let scenario = FailScenario::setup();
    fail::cfg("read-dir", "panic").unwrap();

    do_fallible_work();

    scenario.teardown();
}
```

**Review by dimroc — APPROVED** [review]

**Comment by zekun000** [comment]
to enable it in unit test, you need to specify the required feature for that similar to the doc
```
[[test]]
name = "failpoints"
path = "tests/failpoints/mod.rs"
required-features = ["fail/failpoints"]
```

**Comment by dimroc** [comment]
Is there work to be done to enable it in the smoke test? This is the existing test infra I was referring to.

https://github.com/zekun000/diem/blob/2f445d98d2e1f0c178964be6c50d27b72dda1bf3/testsuite/cluster-test/src/cluster_swarm/configs/validator.yaml#L54-L62

```
# this is only enabled when the binary is compiled with failpoints feature, otherwise no-op
failpoints:
  consensus::process_proposal_msg: 5%return
  consensus::process_sync_info_msg: 5%return
  consensus::process_vote_msg: 5%return
  consensus::process_block_retrieval: 5%return
  consensus::sync_to: 1%return
  consensus::compute: 1%return
  consensus::pull_txn: 1%return
```

**Comment by dimroc** [comment]
This is indeed covered with https://github.com/xli/diem/blob/4d6089c6abb8fcba19ddbffb5bcc290eb69e225d/testsuite/cluster-test/src/cluster_swarm/configs/fullnode.yaml#L45-L57 👍

**Comment by zekun000** [comment]
smoke test enables the feature by default, but there's no configurations of those failpoints (what you paste is for cluster test I think). smoke test uses the binary built here https://github.com/diem/diem/blob/d66e649fd62c92dd7f774ca29e95d0a7b1b9edd7/testsuite/forge/src/backend/local/cargo.rs#L158

**Pull request merged**